php_scrypt.c

```diff
@@ -165,29 +170,31 @@ PHP_FUNCTION(scrypt)
         return;
     }
 
-    /* Clamp & cast them */
-    castError = 0;
-    cryptN = clampAndCast64("N", phpN, &castError, 1);
-    cryptR = clampAndCast32("r", phpR, &castError, 0);
-    cryptP = clampAndCast32("p", phpP, &castError, 0);
+    /* Checks on the parameters */
 
-    if (keyLength < 16) {
-        keyLength = -1;
-        php_error_docref(NULL TSRMLS_CC, E_ERROR, "Key length is too low, must be greater or equal to 16");
-    } else if (keyLength > (powl(2, 32) - 1) * 32) {
-        keyLength = -1;
-        php_error_docref(NULL TSRMLS_CC, E_ERROR, "Key length is too high, must be no more than (2^32 - 1) * 32");
-    }
+	castError = 0;
+	cryptN = clampAndCast64(3, "N", phpN, 1);
+	cryptR = clampAndCast32(4, "r", phpR, 0);
+	cryptP = clampAndCast32(5, "p", phpP, 0);
 
-    /* Return out if we've encountered a error with the input parameters */
-    if (castError > 0 || keyLength < 0) {
-        RETURN_FALSE;
+    /* Return out if we've encountered an exception when casting the arguments */
+    if (EG(exception)) {
+        RETURN_THROWS();
     }
 
-    /* Checks on the parameters */
     if (isPowerOfTwo(cryptN) != 0) {
-      php_error_docref(NULL TSRMLS_CC, E_ERROR, "N parameter must be a power of 2");
-      RETURN_FALSE;
+        zend_throw_error(zend_ce_error, "scrypt(): Argument #3 ($N) must be a power of 2");
+        RETURN_THROWS();
+    }
+
+    if (keyLength < 16) {
+        zend_throw_error(zend_ce_error, "scrypt(): Argument #6 ($key_length) must be greater than or equal to 16");
```

Comment:
Can we bump version requirements to at least PHP 7.0? CI below PHP 5.6 doesn't even run because PHP 5.3-PHP 5.5 is not available anymore, and these changes are incompatible with PHP 5.6 because `Error` is only available since PHP 7.0.